extmod/vfs_blockdev.c

```diff
@@ -46,31 +46,37 @@ void mp_vfs_blockdev_init(mp_vfs_blockdev_t *self, mp_obj_t bdev) {
     }
 }
 
+// Helper function to minimise code size of read/write functions
+static int mp_vfs_blockdev_call_rw(int n_args, mp_obj_t *args, size_t block_num, size_t block_off, size_t len, void *buf) {
+    mp_obj_array_t ar = {{&mp_type_bytearray}, BYTEARRAY_TYPECODE, 0, len, buf};
+    args[2] = MP_OBJ_NEW_SMALL_INT(block_num);
+    args[3] = MP_OBJ_FROM_PTR(&ar);
+    args[4] = MP_OBJ_NEW_SMALL_INT(block_off); // ignored for n_args == 2
+    mp_obj_t ret = mp_call_method_n_kw(n_args, 0, args);
+
+    if (ret == mp_const_none) {
+        return 0;
+    } else {
+        // Block device functions are expected to return 0 on success
+        // and negative integer on errors. Check for positive integer
+        // results as some callers (i.e. littlefs) will produce corrupt
+        // results from these.
+        int i = MP_OBJ_SMALL_INT_VALUE(ret);
+        return i > 0 ? (-MP_EINVAL) : i;
+    }
+}
+
 int mp_vfs_blockdev_read(mp_vfs_blockdev_t *self, size_t block_num, size_t num_blocks, uint8_t *buf) {
     if (self->flags & MP_BLOCKDEV_FLAG_NATIVE) {
         mp_uint_t (*f)(uint8_t *, uint32_t, uint32_t) = (void *)(uintptr_t)self->readblocks[2];
         return f(buf, block_num, num_blocks);
     } else {
-        mp_obj_array_t ar = {{&mp_type_bytearray}, BYTEARRAY_TYPECODE, 0, num_blocks *self->block_size, buf};
-        self->readblocks[2] = MP_OBJ_NEW_SMALL_INT(block_num);
-        self->readblocks[3] = MP_OBJ_FROM_PTR(&ar);
-        mp_call_method_n_kw(2, 0, self->readblocks);
-        // TODO handle error return
-        return 0;
+        return mp_vfs_blockdev_call_rw(2, self->readblocks, block_num, 0, num_blocks * self->block_size, buf);
     }
 }
 
 int mp_vfs_blockdev_read_ext(mp_vfs_blockdev_t *self, size_t block_num, size_t block_off, size_t len, uint8_t *buf) {
-    mp_obj_array_t ar = {{&mp_type_bytearray}, BYTEARRAY_TYPECODE, 0, len, buf};
-    self->readblocks[2] = MP_OBJ_NEW_SMALL_INT(block_num);
-    self->readblocks[3] = MP_OBJ_FROM_PTR(&ar);
-    self->readblocks[4] = MP_OBJ_NEW_SMALL_INT(block_off);
-    mp_obj_t ret = mp_call_method_n_kw(3, 0, self->readblocks);
-    if (ret == mp_const_none) {
-        return 0;
-    } else {
-        return MP_OBJ_SMALL_INT_VALUE(ret);
-    }
+    return mp_vfs_blockdev_call_rw(3, self->readblocks, block_num, block_off, len, buf);
```

Comment:
I had a look at the disassembly, and this is optimised as a tail call on x86-64, but unfortunately not on ARM Thumb2 archs (eg stm32).  That's a shame.

If you put the `n_args` argument at the end of this call, then the generated ARM code is better and smaller because then 4 arguments remain in the same location.  But still it can't do TCO.

Maybe worth putting `n_args` at the end so at least the code is a bit smaller?

Reply:
Oh, well spotted! Have implemented, and as predicted it trimmed a few more bytes off...

I suspect the TCO thing is because the Thumb ports need to spill something to the stack in the top-level function, maybe? Shame it can't optimise that across the board.

Reply:
> I suspect the TCO thing is because the Thumb ports need to spill something to the stack in the top-level function, maybe?

I'm not sure.  It really should be able to adjust the stack so it can do the TCO.  Maybe with -Os it doesn't want to because it makes code size bigger.

Reply:
I was curious: at `-O2` it inlines the static function into each implementation, and `-O1` looks like a less efficient version of `-Os` :shrug:

Reply:
Strange...